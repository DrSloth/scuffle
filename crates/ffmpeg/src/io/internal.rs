use ffmpeg_sys_next::*;
use libc::{c_void, SEEK_CUR, SEEK_END, SEEK_SET};

use crate::error::FfmpegError;
use crate::smart_object::SmartPtr;

const AVERROR_IO: i32 = AVERROR(EIO);

/// Safety: The function must be used with the same type as the one used to
/// generically create the function pointer
pub(crate) unsafe extern "C" fn read_packet<T: std::io::Read>(
    opaque: *mut libc::c_void,
    buf: *mut u8,
    buf_size: i32,
) -> i32 {
    let ret = (*(opaque as *mut T))
        .read(std::slice::from_raw_parts_mut(buf, buf_size as usize))
        .map(|n| n as i32)
        .unwrap_or(AVERROR_IO);

    if ret == 0 {
        return AVERROR_EOF;
    }

    ret
}

/// Safety: The function must be used with the same type as the one used to
/// generically create the function pointer
pub(crate) unsafe extern "C" fn write_packet<T: std::io::Write>(
    opaque: *mut libc::c_void,
    buf: *const u8,
    buf_size: i32,
) -> i32 {
    (*(opaque as *mut T))
        .write(std::slice::from_raw_parts(buf, buf_size as usize))
        .map(|n| n as i32)
        .unwrap_or(AVERROR_IO)
}

/// Safety: The function must be used with the same type as the one used to
/// generically create the function pointer
pub(crate) unsafe extern "C" fn seek<T: std::io::Seek>(opaque: *mut libc::c_void, offset: i64, mut whence: i32) -> i64 {
    let this = &mut *(opaque as *mut T);

    let seek_size = whence & AVSEEK_SIZE != 0;
    if seek_size {
        whence &= !AVSEEK_SIZE;
    }

    let seek_force = whence & AVSEEK_FORCE != 0;
    if seek_force {
        whence &= !AVSEEK_FORCE;
    }

    if seek_size {
        let Ok(pos) = this.stream_position() else {
            return AVERROR_IO as i64;
        };

        let Ok(end) = this.seek(std::io::SeekFrom::End(0)) else {
            return AVERROR_IO as i64;
        };

        if end != pos {
            let Ok(_) = this.seek(std::io::SeekFrom::Start(pos)) else {
                return AVERROR_IO as i64;
            };
        }

        return end as i64;
    }

    let whence = match whence {
        SEEK_SET => std::io::SeekFrom::Start(offset as u64),
        SEEK_CUR => std::io::SeekFrom::Current(offset),
        SEEK_END => std::io::SeekFrom::End(offset),
        _ => return -1,
    };

    match this.seek(whence) {
        Ok(pos) => pos as i64,
        Err(_) => AVERROR_IO as i64,
    }
}

pub(crate) struct Inner<T: Send + Sync> {
    pub(crate) data: Option<Box<T>>,
    pub(crate) context: SmartPtr<AVFormatContext>,
    _io: SmartPtr<AVIOContext>,
}

pub(crate) struct InnerOptions {
    pub(crate) buffer_size: usize,
    pub(crate) read_fn: Option<unsafe extern "C" fn(*mut c_void, *mut u8, i32) -> i32>,
    pub(crate) write_fn: Option<unsafe extern "C" fn(*mut c_void, *const u8, i32) -> i32>,
    pub(crate) seek_fn: Option<unsafe extern "C" fn(*mut c_void, i64, i32) -> i64>,
    pub(crate) output_format: *const AVOutputFormat,
}

impl Default for InnerOptions {
    fn default() -> Self {
        Self {
            buffer_size: 4096,
            read_fn: None,
            write_fn: None,
            seek_fn: None,
            output_format: std::ptr::null(),
        }
    }
}

impl<T: Send + Sync> Inner<T> {
    pub fn new(data: T, options: InnerOptions) -> Result<Self, FfmpegError> {
        // Safety: av_malloc is safe to call
        let buffer = unsafe {
            SmartPtr::wrap_non_null(av_malloc(options.buffer_size), |ptr| {
                // We own this resource so we need to free it
                av_free(*ptr);
                // We clear the old pointer so it doesn't get freed again.
                *ptr = std::ptr::null_mut();
            })
        }
        .ok_or(FfmpegError::Alloc)?;

        let mut data = Box::new(data);

        // Safety: avio_alloc_context is safe to call, and all the function pointers are
        // valid
        let mut io = unsafe {
            SmartPtr::wrap_non_null(
                avio_alloc_context(
                    buffer.as_ptr() as *mut u8,
                    options.buffer_size as i32,
                    if options.write_fn.is_some() { 1 } else { 0 },
                    data.as_mut() as *mut _ as *mut c_void,
                    options.read_fn,
                    options.write_fn,
                    options.seek_fn,
                ),
                |ptr| {
                    // Safety: the pointer is always valid.
                    if let Some(ptr) = ptr.as_mut() {
                        // We need to free the buffer
                        av_free(ptr.buffer as *mut libc::c_void);

                        // We clear the old pointer so it doesn't get freed again.
                        ptr.buffer = std::ptr::null_mut();
                    }

                    avio_context_free(ptr);
                },
            )
        }
        .ok_or(FfmpegError::Alloc)?;

        // The buffer is now owned by the IO context
        buffer.into_inner();

        let mut context = if options.write_fn.is_some() {
            let mut context = unsafe {
                SmartPtr::wrap(std::ptr::null_mut(), |ptr| {
                    // We own this resource so we need to free it
                    avformat_free_context(*ptr);
                    *ptr = std::ptr::null_mut();
                })
            };

            // Safety: avformat_alloc_output_context2 is safe to call
            let ec = unsafe {
                avformat_alloc_output_context2(
                    context.as_mut(),
                    options.output_format,
                    std::ptr::null(),
                    std::ptr::null_mut(),
                )
            };
            if ec != 0 {
                return Err(FfmpegError::Code(ec.into()));
            }

            if context.as_ptr().is_null() {
                return Err(FfmpegError::Alloc);
            }

            context
        } else {
            // Safety: avformat_alloc_context is safe to call
            unsafe {
                SmartPtr::wrap_non_null(avformat_alloc_context(), |ptr| {
                    // We own this resource so we need to free it
                    avformat_free_context(*ptr);
                    *ptr = std::ptr::null_mut();
                })
            }
            .ok_or(FfmpegError::Alloc)?
        };

        // The io context will live as long as the format context
        context.as_deref_mut().expect("Context is null").pb = io.as_mut_ptr();

        Ok(Self {
            data: Some(data),
            context,
            _io: io,
        })
    }
}

impl Inner<()> {
    /// Empty context cannot be used until its initialized and setup correctly
    pub unsafe fn empty() -> Self {
        Self {
            data: Some(Box::new(())),
            context: unsafe {
                SmartPtr::wrap(std::ptr::null_mut(), |ptr| {
                    // We own this resource so we need to free it
                    avformat_free_context(*ptr);
                    *ptr = std::ptr::null_mut();
                })
            },
            _io: unsafe { SmartPtr::wrap(std::ptr::null_mut(), |_| {}) },
        }
    }

    pub fn open_output(path: &str) -> Result<Self, FfmpegError> {
        let path = std::ffi::CString::new(path).expect("Failed to convert path to CString");

        // Safety: avformat_alloc_output_context2 is safe to call
        let mut this = unsafe { Self::empty() };

        // Safety: avformat_alloc_output_context2 is safe to call
        let ec = unsafe {
            avformat_alloc_output_context2(this.context.as_mut(), std::ptr::null(), std::ptr::null(), path.as_ptr())
        };
        if ec != 0 {
            return Err(FfmpegError::Code(ec.into()));
        }

        // We are not moving the pointer so this is safe
        if this.context.as_ptr().is_null() {
            return Err(FfmpegError::Alloc);
        }

        // Safety: avio_open is safe to call
        let ec = unsafe { avio_open(&mut this.context.as_deref_mut_except().pb, path.as_ptr(), AVIO_FLAG_WRITE) };

        if ec != 0 {
            return Err(FfmpegError::Code(ec.into()));
        }

        this.context.set_destructor(|ptr| unsafe {
            // We own this resource so we need to free it
            avio_closep(&mut (**ptr).pb);
            avformat_free_context(*ptr);
            *ptr = std::ptr::null_mut();
        });

        Ok(this)
    }
}

#[cfg(test)]
#[cfg_attr(all(test, coverage_nightly), coverage(off))]
mod tests {
    use std::ffi::CString;
    use std::io::Cursor;
    use std::sync::atomic::{AtomicUsize, Ordering};
    use std::sync::Once;

    use ffmpeg_sys_next::{av_guess_format, AVSEEK_FORCE};
    use libc::{c_void, SEEK_CUR, SEEK_END};
    use tempfile::Builder;

    use crate::error::FfmpegError;
    use crate::io::internal::{read_packet, seek, write_packet, Inner, InnerOptions, AVERROR_EOF};

    #[test]
    fn test_read_packet_eof() {
        let mut data: Cursor<Vec<u8>> = Cursor::new(vec![]);
        let mut buf = [0u8; 10];

        unsafe {
            let result =
                read_packet::<Cursor<Vec<u8>>>((&raw mut data) as *mut libc::c_void, buf.as_mut_ptr(), buf.len() as i32);

            assert_eq!(result, AVERROR_EOF);
        }
    }

    #[test]
    fn test_write_packet_success() {
        let mut data = Cursor::new(vec![0u8; 10]);
        let buf = [1u8, 2, 3, 4, 5];

        unsafe {
            let result =
                write_packet::<Cursor<Vec<u8>>>((&raw mut data) as *mut _ as *mut c_void, buf.as_ptr(), buf.len() as i32);
            assert_eq!(result, buf.len() as i32);

            let written_data = data.get_ref();
            assert_eq!(&written_data[..buf.len()], &buf);
        }
    }

    #[test]
    fn test_seek_force() {
        let mut cursor = Cursor::new(vec![0u8; 100]);
        let opaque = &mut cursor as *mut _ as *mut c_void;
        assert_eq!(cursor.position(), 0);
        let offset = 10;
        let mut whence = SEEK_CUR | AVSEEK_FORCE;
        let result = unsafe { seek::<Cursor<Vec<u8>>>(opaque, offset, whence) };

        assert_eq!(result, { offset });
        whence &= !AVSEEK_FORCE;
        assert_eq!(whence, SEEK_CUR);
        assert_eq!(cursor.position(), offset as u64);
    }

    #[test]
    fn test_seek_seek_end() {
        let mut cursor = Cursor::new(vec![0u8; 100]);
        let opaque = &mut cursor as *mut _ as *mut libc::c_void;
        let offset = -10;
        let whence = SEEK_END;
        let result = unsafe { seek::<Cursor<Vec<u8>>>(opaque, offset, whence) };

        assert_eq!(result, 90);
        assert_eq!(cursor.position(), 90);
    }

    #[test]
    fn test_seek_invalid_whence() {
        let mut cursor = Cursor::new(vec![0u8; 100]);
        let opaque = &mut cursor as *mut _ as *mut libc::c_void;
        let result = unsafe { seek::<Cursor<Vec<u8>>>(opaque, 0, 999) };

        assert_eq!(result, -1);
        assert_eq!(cursor.position(), 0);
    }

    #[test]
    fn test_avformat_alloc_output_context2_error() {
        static BUF_SIZE_TRACKER: AtomicUsize = AtomicUsize::new(0);
        static CALL_COUNT: AtomicUsize = AtomicUsize::new(0);
        static INIT: Once = Once::new();

        INIT.call_once(|| {
            BUF_SIZE_TRACKER.store(0, Ordering::SeqCst);
            CALL_COUNT.store(0, Ordering::SeqCst);
        });

        unsafe extern "C" fn dummy_write_fn(_opaque: *mut libc::c_void, _buf: *const u8, _buf_size: i32) -> i32 {
            CALL_COUNT.fetch_add(1, Ordering::SeqCst);
            BUF_SIZE_TRACKER.store(_buf_size as usize, Ordering::SeqCst);
            0 // simulate success
        }

        let invalid_format = CString::new("invalid_format").expect("Failed to create CString");
        let options = InnerOptions {
            buffer_size: 4096,
            write_fn: Some(dummy_write_fn),
            output_format: unsafe { av_guess_format(invalid_format.as_ptr(), std::ptr::null(), std::ptr::null()) },
            ..Default::default()
        };
        let data = ();
        let result = Inner::new(data, options);

        assert!(result.is_err(), "Expected an error but got Ok");

        let call_count = CALL_COUNT.load(Ordering::SeqCst);
        assert_eq!(call_count, 0, "Expected dummy_write_fn to not be called.");

        if let Err(error) = result {
            match error {
                FfmpegError::Code(_) => {
                    eprintln!("Expected avformat_alloc_output_context2 error occurred.");
                }
                _ => panic!("Unexpected error variant: {:?}", error),
            }
        }
    }

    #[test]
    fn test_open_output_valid_path() {
        let temp_file = Builder::new()
            .suffix(".mp4")
            .tempfile()
            .expect("Failed to create a temporary file");
        let test_path = temp_file.path();
        let result = Inner::open_output(test_path.to_str().unwrap());

        assert!(result.is_ok(), "Expected success but got error");
    }

    #[test]
    fn test_open_output_invalid_path() {
        let test_path = "";
        let result = Inner::open_output(test_path);

        assert!(result.is_err(), "Expected Err, got Ok");
    }

    #[test]
    fn test_open_output_avformat_alloc_error() {
        let test_path = tempfile::tempdir().unwrap().path().join("restricted_output.mp4");
        let test_path_str = test_path.to_str().unwrap();
        let result = Inner::open_output(test_path_str);
        if let Err(error) = &result {
            eprintln!("Function returned an error: {:?}", error);
        }

        assert!(
            matches!(result, Err(FfmpegError::Code(_))),
            "Expected FfmpegError::Code but received a different error."
        );
    }
}
