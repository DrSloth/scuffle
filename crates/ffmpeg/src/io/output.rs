use ffmpeg_sys_next::*;

use super::internal::{seek, write_packet, Inner, InnerOptions};
use crate::consts::DEFAULT_BUFFER_SIZE;
use crate::dict::Dictionary;
use crate::error::FfmpegError;
use crate::packet::Packet;
use crate::stream::Stream;

#[derive(Debug, Clone)]
pub struct OutputOptions<'a> {
    pub buffer_size: usize,
    pub format_name: Option<&'a str>,
    pub format_mime_type: Option<&'a str>,
    pub format_ffi: *const AVOutputFormat,
}

impl<'a> OutputOptions<'a> {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn buffer_size(mut self, buffer_size: usize) -> Self {
        self.buffer_size = buffer_size;
        self
    }

    pub fn format_name(mut self, format_name: &'a str) -> Self {
        self.format_name = format_name.into();
        self
    }

    pub fn format_mime_type(mut self, format_mime_type: &'a str) -> Self {
        self.format_mime_type = format_mime_type.into();
        self
    }

    pub fn format_ffi(mut self, format_ffi: *const AVOutputFormat) -> Self {
        self.format_ffi = format_ffi;
        self
    }

    fn get_format_ffi(&self) -> Result<*const AVOutputFormat, FfmpegError> {
        if !self.format_ffi.is_null() {
            return Ok(self.format_ffi);
        }

        if self.format_name.is_none() && self.format_mime_type.is_none() {
            return Err(FfmpegError::Arguments(
                "format_ffi, format_name and format_mime_type cannot all be unset",
            ));
        }

        let c_format_name = self.format_name.map(|s| std::ffi::CString::new(s).unwrap());
        let c_format_mime_type = self.format_mime_type.map(|s| std::ffi::CString::new(s).unwrap());
        let c_format_name_ptr = c_format_name.as_ref().map(|s| s.as_ptr()).unwrap_or(std::ptr::null());
        let c_format_mime_type_ptr = c_format_mime_type.as_ref().map(|s| s.as_ptr()).unwrap_or(std::ptr::null());

        // Safety: `av_guess_format` is safe to call with null pointers.
        let output_format = unsafe { av_guess_format(c_format_name_ptr, std::ptr::null(), c_format_mime_type_ptr) };

        if output_format.is_null() {
            return Err(FfmpegError::Arguments("could not determine output format"));
        }

        Ok(output_format)
    }
}

impl Default for OutputOptions<'_> {
    fn default() -> Self {
        Self {
            buffer_size: DEFAULT_BUFFER_SIZE,
            format_name: None,
            format_mime_type: None,
            format_ffi: std::ptr::null(),
        }
    }
}

pub struct Output<T: Send + Sync> {
    inner: Inner<T>,
    witten_header: bool,
}

/// Safety: `T` must be `Send` and `Sync`.
unsafe impl<T: Send + Sync> Send for Output<T> {}

impl<T: Send + Sync> Output<T> {
    pub fn into_inner(mut self) -> T {
        *(self.inner.data.take().unwrap())
    }
}

impl<T: std::io::Write + Send + Sync> Output<T> {
    pub fn new(input: T, options: OutputOptions) -> Result<Self, FfmpegError> {
        let output_format = options.get_format_ffi()?;

        Ok(Self {
            inner: Inner::new(
                input,
                InnerOptions {
                    buffer_size: options.buffer_size,
                    write_fn: Some(write_packet::<T>),
                    output_format,
                    ..Default::default()
                },
            )?,
            witten_header: false,
        })
    }

    pub fn seekable(input: T, options: OutputOptions) -> Result<Self, FfmpegError>
    where
        T: std::io::Seek,
    {
        let output_format = options.get_format_ffi()?;

        Ok(Self {
            inner: Inner::new(
                input,
                InnerOptions {
                    buffer_size: options.buffer_size,
                    write_fn: Some(write_packet::<T>),
                    seek_fn: Some(seek::<T>),
                    output_format,
                    ..Default::default()
                },
            )?,
            witten_header: false,
        })
    }
}

impl<T: Send + Sync> Output<T> {
    pub fn set_metadata(&mut self, metadata: Dictionary) {
        unsafe {
            // This frees the old metadata
            Dictionary::from_ptr_owned(self.inner.context.as_deref_mut_except().metadata);
            self.inner.context.as_deref_mut_except().metadata = metadata.into_ptr();
        };
    }

    pub fn as_ptr(&self) -> *const AVFormatContext {
        self.inner.context.as_ptr()
    }

    pub fn as_mut_ptr(&mut self) -> *mut AVFormatContext {
        self.inner.context.as_mut_ptr()
    }

    pub fn add_stream(&mut self, codec: Option<*const AVCodec>) -> Option<Stream<'_>> {
        // Safety: `avformat_new_stream` is safe to call.
        let stream = unsafe { avformat_new_stream(self.as_mut_ptr(), codec.unwrap_or_else(std::ptr::null)) };
        if stream.is_null() {
            None
        } else {
            // Safety: 'stream' is a valid pointer here.
            unsafe {
                let stream = &mut *stream;
                stream.id = self.inner.context.as_deref_except().nb_streams as i32 - 1;
                Some(Stream::new(stream, self.inner.context.as_deref_except()))
            }
        }
    }

    pub fn copy_stream<'a>(&'a mut self, stream: &Stream<'_>) -> Option<Stream<'a>> {
        let codec_param = stream.codec_parameters()?;

        // Safety: `avformat_new_stream` is safe to call.
        let out_stream = unsafe { avformat_new_stream(self.as_mut_ptr(), std::ptr::null()) };
        if out_stream.is_null() {
            None
        } else {
            // Safety: 'out_stream', 'codec_param' and 'context' are valid pointers here.
            unsafe {
                let out_stream = &mut *out_stream;

                // Safety: `avcodec_parameters_copy` is safe to call.
                avcodec_parameters_copy(out_stream.codecpar, codec_param);
                out_stream.id = self.inner.context.as_deref_except().nb_streams as i32 - 1;
                let mut out_stream = Stream::new(out_stream, self.inner.context.as_deref_except());
                out_stream.set_time_base(stream.time_base());
                out_stream.set_start_time(stream.start_time());
                out_stream.set_duration(stream.duration());

                Some(out_stream)
            }
        }
    }

    pub fn write_header(&mut self) -> Result<(), FfmpegError> {
        if self.witten_header {
            return Err(FfmpegError::Arguments("header already written"));
        }

        // Safety: `avformat_write_header` is safe to call, if the header has not been
        // written yet.
        unsafe {
            match avformat_write_header(self.as_mut_ptr(), std::ptr::null_mut()) {
                0 => Ok(()),
                e => Err(FfmpegError::Code(e.into())),
            }
        }?;

        self.witten_header = true;
        Ok(())
    }

    pub fn write_header_with_options(&mut self, options: &mut Dictionary) -> Result<(), FfmpegError> {
        if self.witten_header {
            return Err(FfmpegError::Arguments("header already written"));
        }

        // Safety: `avformat_write_header` is safe to call, if the header has not been
        // written yet.
        unsafe {
            match avformat_write_header(self.as_mut_ptr(), options.as_mut_ptr_ref()) {
                0 => Ok(()),
                e => Err(FfmpegError::Code(e.into())),
            }
        }?;

        self.witten_header = true;
        Ok(())
    }

    pub fn write_trailer(&mut self) -> Result<(), FfmpegError> {
        if !self.witten_header {
            return Err(FfmpegError::Arguments("header not written"));
        }

        // Safety: `av_write_trailer` is safe to call, once the header has been written.
        unsafe {
            match av_write_trailer(self.as_mut_ptr()) {
                n if n >= 0 => Ok(()),
                e => Err(FfmpegError::Code(e.into())),
            }
        }
    }

    pub fn write_interleaved_packet(&mut self, mut packet: Packet) -> Result<(), FfmpegError> {
        if !self.witten_header {
            return Err(FfmpegError::Arguments("header not written"));
        }

        // Safety: `av_interleaved_write_frame` is safe to call, once the header has
        // been written.
        unsafe {
            match av_interleaved_write_frame(self.as_mut_ptr(), packet.as_mut_ptr()) {
                0 => Ok(()),
                e => Err(FfmpegError::Code(e.into())),
            }
        }
    }

    pub fn write_packet(&mut self, packet: &Packet) -> Result<(), FfmpegError> {
        if !self.witten_header {
            return Err(FfmpegError::Arguments("header not written"));
        }

        // Safety: `av_write_frame` is safe to call, once the header has been written.
        unsafe {
            match av_write_frame(self.as_mut_ptr(), packet.as_ptr() as *mut _) {
                0 => Ok(()),
                e => Err(FfmpegError::Code(e.into())),
            }
        }
    }

    pub fn flags(&self) -> i32 {
        self.inner.context.as_deref_except().flags
    }
}

impl Output<()> {
    pub fn open(path: &str) -> Result<Self, FfmpegError> {
        Ok(Self {
            inner: Inner::open_output(path)?,
            witten_header: false,
        })
    }
}

#[cfg(test)]
#[cfg_attr(all(test, coverage_nightly), coverage(off))]
mod tests {
    use std::ffi::CString;
    use std::io::Cursor;
    use std::ptr;

    use tempfile::Builder;

    use crate::consts::DEFAULT_BUFFER_SIZE;
    use crate::dict::Dictionary;
    use crate::error::FfmpegError;
    use crate::io::output::{AVCodec, AVRational, AVFMT_FLAG_AUTO_BSF};
    use crate::io::{Output, OutputOptions};

    #[test]
    fn test_output_options_default() {
        let options = OutputOptions::default();

        assert_eq!(options.buffer_size, DEFAULT_BUFFER_SIZE);
        assert!(options.format_name.is_none());
        assert!(options.format_mime_type.is_none());
        assert!(options.format_ffi.is_null());
    }

    #[test]
    fn test_output_options_new() {
        let options = OutputOptions::new();

        assert_eq!(options.buffer_size, DEFAULT_BUFFER_SIZE);
        assert!(options.format_name.is_none());
        assert!(options.format_mime_type.is_none());
        assert!(options.format_ffi.is_null());
    }

    #[test]
    fn test_output_options_custom_values() {
        let options = OutputOptions::default()
            .buffer_size(4096)
            .format_name("mp4")
            .format_mime_type("video/mp4");

        assert_eq!(options.buffer_size, 4096);
        assert_eq!(options.format_name, Some("mp4"));
        assert_eq!(options.format_mime_type, Some("video/mp4"));
    }

    #[test]
    fn test_output_options_get_format_ffi() {
        let options = OutputOptions::default().format_name("mp4");
        let format_ffi = options.get_format_ffi();
        assert!(format_ffi.is_ok());
        assert!(!format_ffi.unwrap().is_null());
    }

    #[test]
    fn test_output_options_get_format_ffi_null() {
        let format_name = CString::new("mp4").unwrap();
        let format_mime_type = CString::new("").unwrap();
        let format_ptr = unsafe { ffmpeg_sys_next::av_guess_format(format_name.as_ptr(), ptr::null(), format_mime_type.as_ptr()) };

        assert!(!format_ptr.is_null(), "Failed to retrieve AVOutputFormat for the given format name");

        let output_options = OutputOptions::new().format_ffi(format_ptr);
        let result = output_options.get_format_ffi();

        assert!(result.is_ok(), "Expected Ok result, got Err instead");
        assert_eq!(
            result.unwrap(),
            format_ptr,
            "Expected format_ffi pointer to match the retrieved format"
        );
    }

    #[test]
    fn test_output_options_get_format_ffi_unset_error() {
        let options = OutputOptions::default();
        let format_ffi = options.get_format_ffi();
        assert!(format_ffi.is_err());
    }

    #[test]
    fn test_output_options_get_format_ffi_output_format_error() {
        let options = OutputOptions::default().format_name("unknown_format");
        let format_ffi_result = options.get_format_ffi();

        assert!(format_ffi_result.is_err());
        match format_ffi_result {
            Err(FfmpegError::Arguments(msg)) => {
                assert_eq!(msg, "could not determine output format");
            }
            _ => panic!("Expected FfmpegError::Arguments"),
        }
    }

    #[test]
    fn test_output_into_inner() {
        let data = Cursor::new(Vec::with_capacity(1024));
        let options = OutputOptions::default().format_name("mp4");
        let output = Output::new(data, options).expect("Failed to create Output");
        let inner_data = output.into_inner();

        assert!(inner_data.get_ref().is_empty());
        let buffer = inner_data.into_inner();
        assert_eq!(buffer.capacity(), 1024);
    }

    #[test]
    fn test_output_new() {
        let data = Cursor::new(Vec::new());
        let options = OutputOptions::default().format_name("mp4");
        let output = Output::new(data, options);

        assert!(output.is_ok());
    }

    #[test]
    fn test_output_seekable() {
        let data = Cursor::new(Vec::new());
        let options = OutputOptions::default().format_name("mp4");
        let output = Output::seekable(data, options);

        assert!(output.is_ok());
    }

    #[test]
    fn test_output_set_metadata() {
        let data = Cursor::new(Vec::new());
        let options = OutputOptions::default().format_name("mp4");
        let mut output = Output::new(data, options).unwrap();
        let metadata = Dictionary::new();
        output.set_metadata(metadata);

        assert!(!output.as_ptr().is_null());
    }

    #[test]
    fn test_output_as_mut_ptr() {
        let data = Cursor::new(Vec::new());
        let options = OutputOptions::default().format_name("mp4");
        let mut output = Output::new(data, options).expect("Failed to create Output");
        let context_ptr = output.as_mut_ptr();

        assert!(!context_ptr.is_null(), "Expected non-null pointer from as_mut_ptr");
    }

    #[test]
    fn test_add_stream_with_valid_codec() {
        let data = Cursor::new(Vec::new());
        let options = OutputOptions::default().format_name("mp4");
        let mut output = Output::new(data, options).expect("Failed to create Output");
        let dummy_codec: *const AVCodec = 0x1234 as *const AVCodec;
        let stream = output.add_stream(Some(dummy_codec));

        assert!(stream.is_some(), "Expected a valid Stream to be added");
    }

    #[test]
    fn test_copy_stream() {
        let data = Cursor::new(Vec::new());
        let options = OutputOptions::default().format_name("mp4");
        let mut output = Output::new(data, options).expect("Failed to create Output");

        // create new output to prevent double mut borrow
        let data = Cursor::new(Vec::new());
        let options = OutputOptions::default().format_name("mp4");
        let mut output_two = Output::new(data, options).expect("Failed to create Output");

        let dummy_codec: *const AVCodec = 0x1234 as *const AVCodec;
        let mut source_stream = output.add_stream(Some(dummy_codec)).expect("Failed to add source stream");

        source_stream.set_time_base(AVRational { num: 1, den: 25 });
        source_stream.set_start_time(Some(1000));
        source_stream.set_duration(Some(500));
        let copied_stream = output_two.copy_stream(&source_stream).expect("Failed to copy the stream");

        assert_eq!(copied_stream.index(), source_stream.index(), "Stream indices should match");
        assert_eq!(copied_stream.id(), source_stream.id(), "Stream IDs should match");
        assert_eq!(
            copied_stream.time_base(),
            source_stream.time_base(),
            "Time bases should match"
        );
        assert_eq!(
            copied_stream.start_time(),
            source_stream.start_time(),
            "Start times should match"
        );
        assert_eq!(copied_stream.duration(), source_stream.duration(), "Durations should match");
        assert_eq!(
            copied_stream.duration(),
            source_stream.duration(),
            "Durations should match"
        );
        assert!(!copied_stream.as_ptr().is_null(), "Copied stream pointer should not be null");
    }

    #[test]
    fn test_output_flags() {
        let data = Cursor::new(Vec::new());
        let options = OutputOptions::default().format_name("mp4");
        let output = Output::new(data, options).expect("Failed to create Output");
        let flags = output.flags();

        assert_eq!(flags, AVFMT_FLAG_AUTO_BSF, "Expected default flag to be AVFMT_FLAG_AUTO_BSF");
    }

    #[test]
    fn test_output_open() {
        let temp_file = Builder::new()
            .suffix(".mp4")
            .tempfile()
            .expect("Failed to create a temporary file");
        let temp_path = temp_file.path();
        let output = Output::open(temp_path.to_str().unwrap());

        assert!(output.is_ok(), "Expected Output::open to succeed");
        std::fs::remove_file(temp_path).expect("Failed to remove temporary file");
    }
}
