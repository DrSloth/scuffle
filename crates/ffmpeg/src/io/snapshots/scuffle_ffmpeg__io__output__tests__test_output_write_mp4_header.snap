---
source: crates/ffmpeg/src/io/output.rs
expression: get_boxes!(output)
snapshot_kind: text
---
[
    Ftyp(
        Ftyp {
            header: BoxHeader {
                box_type: b"ftyp",
            },
            major_brand: Unknown(
                [
                    105,
                    115,
                    111,
                    109,
                ],
            ),
            minor_version: 512,
            compatible_brands: [
                Unknown(
                    [
                        105,
                        115,
                        111,
                        109,
                    ],
                ),
                Unknown(
                    [
                        105,
                        115,
                        111,
                        50,
                    ],
                ),
                Avc1,
                Mp41,
            ],
        },
    ),
    Unknown(
        (
            BoxHeader {
                box_type: b"free",
            },
            b"",
        ),
    ),
    Mdat(
        Mdat {
            header: BoxHeader {
                box_type: b"mdat",
            },
            data: [
                b"\xe3\xb0\xc4B\x98\xfc\x1c\x14\x9a\xfb\xf4\xc8\x99o\xb9$'\xaeA\xe4d\x9b\x93L\xa4\x95\x99\x1bxR\xb8U",
            ],
        },
    ),
]
